src/Hopper/Internal/LoweredCore/ANF.hs

```diff
@@ -0,0 +1,477 @@
+  nextRef.refId %= succ
+  return curr
+
+-- | In converting a 'Term' to ANF, specifies whether the allocated value or
+-- result of (prim/fun/thunk) application should be bound to an 'AnfLet'
+-- corresponding to an existing ('Let') 'TermBinding' or a new 'AnfBinding'.
+data Binding
+  = TermBinding
+  | AnfBinding [AnfRef] -- A list because we use unboxed tuples and 2D binders.
+  deriving (Eq, Show)
+
+-- | Bumps the provided source variable by the number of intermediate lets that
+-- have been introduced since the source binder this variable points to.
+translateTermVar :: Variable -> BindingStack -> Variable
+translateTermVar var@(GlobalVarSym _) _ = var
+translateTermVar var@(LocalVar lnv) stack =
+  case mIndirections of
+    Nothing ->
+      adjust displacement var
+    Just indirections ->
+      adjust (displacement + depth) $ indirections V.! slot
+
+  where
+    depth = lnv ^. lnDepth
+    slot  = lnv ^. lnSlot.slotIndex.to fromIntegral
+
+    levels :: [BindingLevel]
+    levels  = take (fromIntegral $ depth + 1) stack
+
+    -- Calculate the displacement sum (from the number of let-introductions) and
+    -- grab the last (possible) indirection in a single pass:
+    (displacement, mIndirections) = (getSum *** join . getLast) . mconcat $
+      (Sum . _levelIntros &&& Last . Just . view levelIndirections) <$> levels
+
+    adjust :: Word32 -> Variable -> Variable
+    adjust offset = localNameless.lnDepth +~ offset
+
+-- | Updates the top of the 'BindingStack' to account for the existing
+-- (pre-translated) 'Term' 'Variable's.
+trackVariables :: Binding
+               -> [Variable]
+               -- ^ Existing source variables that've already been translated to
+               -- point an appropriate number of binders up in ANF land.
+               -> BindingStack
+               -> BindingStack
+trackVariables (AnfBinding refs) vars stack =
+  stack & _head.levelRefs %~ addRefs
+  where
+    addRefs m = foldl' (flip $ uncurry Map.insert) m $ zip refs vars
+trackVariables TermBinding vars stack =
+  emptyIndirectionLevel vars : stack
+
+-- | Updates the top of the 'BindingStack' to open a new 'AnfRef' for an
+-- introduced 'AnfLet', or adds a new 'BindingLevel' for an existing 'Term'
+-- 'Let'.
+trackBinding :: Binding -> BindingStack -> BindingStack
+trackBinding (AnfBinding refs) stack = stack & _head %~ updateLevel
+  where
+    updateLevel = (levelRefs                              %~ addRefs)
+                . (levelRefs.mapped.localNameless.lnDepth +~ 1)
+                . (levelIntros                            +~ 1)
+    addRefs m = foldl' (\m' ref -> Map.insert ref v0 m') m refs
+trackBinding TermBinding stack = emptyLevel:stack
+
+-- | Stops tracking the provided 'AnfRef's in the top 'BindingLevel' of
+-- 'BindingStack'.
+dropRefs :: Foldable t => t AnfRef -> BindingStack -> BindingStack
+dropRefs refs stack = stack & _head.levelRefs %~ deleteRefs
+  where
+    deleteRefs m = foldl' (flip Map.delete) m refs
+
```

Comment:
leaving this one as-is because there's no `Map` function for removing a list of keys